backends/tc/introspection.cpp

```diff
@@ -0,0 +1,299 @@
+}
+
+cstring IntrospectionGenerator::externalName(const IR::IDeclaration *declaration) {
+    cstring name = declaration->externalName();
+    if (name.startsWith(".")) name = name.substr(1);
+    auto Name = name.replace('.', '/');
+    return Name;
+}
+
+void IntrospectionGenerator::collectTableInfo() {
+    for (auto table : tcPipeline->tableDefs) {
+        auto tableInfo = new struct tableAttributes();
+        tableInfo->id = table->tableID;
+        tableInfo->name = table->controlName + "/" + table->tableName;
+        tableInfo->tentries = table->tableEntriesCount;
+        tableInfo->numMask = table->numMask;
+        if (table->keySize != 0) {
+            tableInfo->keysize = table->keySize;
+        }
+        // Key field information collection
+        unsigned int i = 1;
+        const IR::P4Table *p4table = nullptr;
+        p4table = p4tables[table->tableName];
+        if (p4table != nullptr) {
+            auto key = p4table->getKey();
+            if (key != nullptr && key->keyElements.size()) {
+                for (auto k : key->keyElements) {
+                    auto keyField = new struct keyFieldAttributes();
+                    keyField->id = i++;
+                    auto keyExp = k->expression;
+                    keyField->name = keyExp->toString();
+                    keyField->matchType = k->matchType->toString();
+                    auto keyExpType = typeMap->getType(keyExp);
+                    auto widthBits = keyExpType->width_bits();
+                    keyField->type = "bit" + Util::toString(widthBits);
+                    auto keyAnno = k->getAnnotations()->annotations;
+                    for (auto anno : keyAnno) {
+                        if (anno->name == "tc_type") {
+                            auto expressionList = anno->expr;
+                            for (auto expr : expressionList) {
+                                if (auto typeLiteral = expr->to<IR::StringLiteral>()) {
+                                    auto val = checkValidTcType(typeLiteral);
+                                    if (val != nullptr) {
+                                        keyField->type = val;
+                                    } else {
+                                        ::error(ErrorType::ERR_INVALID,
+                                                "tc_type annotation cannot have '%1%' as value",
+                                                expr);
+                                    }
+                                } else {
+                                    ::error(ErrorType::ERR_INVALID,
+                                            "tc_type annotation cannot have '%1%' as value", expr);
+                                }
+                            }
+                        }
+                        if (anno->name == "name") {
+                            auto expressionList = anno->expr;
+                            for (auto expr : expressionList) {
```

Comment:
Can there be multiple elements in the name annotation expression list?

Reply:
List can have only one value here. Updated the code